# 10.错误处理与异常

## 10.1 关于错误处理

* **数据错误**
  * 在数字金额的一栏填上了商品名
* **系统错误**
  * 硬盘故障等明显的错误
* **程序错误**
  * 因调用了不存在的方法、弄错参数值或算法错误以及程序本身的缺陷产生的错误

对错误的处理：

* **排除错误的原因**
* **忽略错误**
  * 程序有一些无伤大雅错误时可以忽略
* **恢复错误发生前的状态**
  * 向用户提示错误，并指导用户如何操作
* **重试一次**
  * 曾经执行失败的程序，过一段时间后再重新执行可能就会成功
* **终止程序**
  * 只是一个人用的小程序，也许本来就没有必要做错误处理

对于可预期的错误，要注意：

* **是否破坏了输入的数据，特别是人工制作的数据**
* **是否可以对错误的内容及其原因做出相应的提示**

## 10.2 异常处理

错误：

```text
文件名：行号：in`方法名'：错误信息（异常类名）
    from 文件名：行号：in 方法名
    ···
```

ruby中的异常处理：

```text
begin
    a()
    b()
    c()
rescue
    错误处理
end
# a(),b(),c()中发生异常
```

## 10.3 异常处理的写法

Ruby中使用begin···rescue\`\`\`end语句描述一场处理。

```text
begin
    可能会发生异常的处理
rescue =>引用异常对象的变量
    发生异常时的处理
end
#异常发生时被自动赋值的变量
# $!  ： 最后发生的异常（异常对象）
# $@  ： 最后发生的异常的位置信息
```

异常对象的方法

| 方法名 | 意义 |
| :--- | :--- |
| class | 异常的种类 |
| message | 异常信息 |
| backtrace | 异常发生的位置信息（$@与$!.backtrace等价） |

```ruby
ltotal = 0
wtotal = 0
ctotal = 0
ARGV.each do |file|
	begin
		input = File.open(file)
		l = 0
		w = 0
		c = 0
		input.each_line do |line|
			l += l
			c +=line.size
			line.sub!(/^\s+/," ")
			ary = line.split(/\s+/)
			w +=ary.size
		end
		input.close
		printf("%8d %8d %8d %s\n",l,w,c,file)
		ltotal +=l
		wtotal +=w
		ctotal +=c
	rescue =>ex
		print ex.message,"\n"
	end
end
printf("%8d %8d %8d %s\n",ltotal,wtotal,ctotal,"total")
```

![](../.gitbook/assets/image%20%28140%29.png)

